plugin to rewrite markdown links

Aierie opened the pull request:
https://github.com/orgs/tauri-apps/projects/18/views/1?pane=issue&itemId=19155863

This plugin modifies relative links to other markdown files by: 

- removing `.md` file extensions
- resolving inconsistencies between relative linking in filesystems vs in urls

It assumes that we're using [Netlify pretty url configuration that applies trailing slashes](https://docs.netlify.com/routing/redirects/redirect-options/#trailing-slash).

Aierie committed:
plugin to rewrite markdown links
- assume trailing slash urls

Aierie commented:
Unsure where to put these, but here are tests for the `transformLink` function (along with a JS version of it), which can be run with the built-in test runner on node 18:

```
import test from "node:test";
import { strict as assert } from "node:assert";

function removeIndexFromEnd(link) {
  return link.replace(/(\/|^)index$/, "");
}

function removeExplicitCurrentDir(link) {
  return link.replace(/^.\//, "");
}

// Not handling full paths since they don"t really make sense within individual markdown files
function transformLink(link, currentFileIsIndex) {
  link = link.trim();
  if (
    /^\/\//.test(link) ||
    /^http(s)?:\/\//.test(link) ||
    !/\.md(#[^/]+)?$/.test(link)
  ) {
    return link;
  }

  let [pathToMdFile, afterPoundSign] = link.split(/(?<=\.md)#/);
  let fragment = afterPoundSign ? `#${afterPoundSign}` : "";
  let pathToPage = pathToMdFile.replace(/\.md$/, "");

  let urlPath = removeIndexFromEnd(removeExplicitCurrentDir(pathToPage));
  if (!currentFileIsIndex) {
    // Because of the trailing slashes in urls, we always want to move one level up if we aren't on an index page
    urlPath = ".." + (urlPath ? `/${urlPath}` : "");
  }

  if (urlPath === "") {
    return fragment || "#";
  } else {
    // Add the trailing slash at the end, then the fragment if necessary
    return urlPath + "/" + fragment;
  }
}

// Assume trailing slash in all urls
test("base/dir_1/nonindex.md -> base/dir_1/nonindex_2.md", () => {
  let currentFile = "base/dir_1/nonindex.md"; // url is base/dir_1/nonindex/, destination is base/dir_1/nonindex_2/
  assert.equal(
    transformLink("./nonindex_2.md", currentFile.endsWith("/index.md")),
    "../nonindex_2/"
  );
  assert.equal(
    transformLink("nonindex_2.md", currentFile.endsWith("/index.md")),
    "../nonindex_2/"
  );
  assert.equal(
    transformLink("./nonindex_2.md#id", currentFile.endsWith("/index.md")),
    "../nonindex_2/#id"
  );
  assert.equal(
    transformLink("nonindex_2.md#id", currentFile.endsWith("/index.md")),
    "../nonindex_2/#id"
  );
});
test("base/dir_1/nonindex.md -> base/dir_2/nonindex.md", () => {
  let currentFile = "base/dir_1/nonindex.md"; // url is base/dir_1/nonindex/, destination is base/dir_2/nonindex/
  assert.equal(
    transformLink("../dir_2/nonindex.md", currentFile.endsWith("/index.md")),
    "../../dir_2/nonindex/"
  );
  assert.equal(
    transformLink("../dir_2/nonindex.md#id", currentFile.endsWith("/index.md")),
    "../../dir_2/nonindex/#id"
  );
});
test("base/dir_1/nonindex.md -> base/dir_1/index.md", () => {
  let currentFile = "base/dir_1/nonindex.md"; // url is base/dir_1/nonindex/, destination is base/dir_1/
  assert.equal(
    transformLink("./index.md", currentFile.endsWith("/index.md")),
    "../"
  );
  assert.equal(
    transformLink("index.md", currentFile.endsWith("/index.md")),
    "../"
  );
  assert.equal(
    transformLink("./index.md#id", currentFile.endsWith("/index.md")),
    "../#id"
  );
  assert.equal(
    transformLink("index.md#id", currentFile.endsWith("/index.md")),
    "../#id"
  );
});
test("base/dir_1/nonindex.md -> base/dir_2/index.md", () => {
  let currentFile = "base/dir_1/nonindex.md"; // url is base/dir_1/nonindex/, destination is base/dir_2/
  assert.equal(
    transformLink("../dir_2/index.md", currentFile.endsWith("/index.md")),
    "../../dir_2/"
  );
  assert.equal(
    transformLink("../dir_2/index.md#id", currentFile.endsWith("/index.md")),
    "../../dir_2/#id"
  );
});
test("base/dir_1/index.md -> base/dir_1/nonindex.md", () => {
  let currentFile = "base/dir_1/index.md"; // url is base/dir_1/, destination is base/dir_1/nonindex/
  assert.equal(
    transformLink("./nonindex.md", currentFile.endsWith("/index.md")),
    "nonindex/"
  );
  assert.equal(
    transformLink("nonindex.md", currentFile.endsWith("/index.md")),
    "nonindex/"
  );
  assert.equal(
    transformLink("./nonindex.md#id", currentFile.endsWith("/index.md")),
    "nonindex/#id"
  );
  assert.equal(
    transformLink("nonindex.md#id", currentFile.endsWith("/index.md")),
    "nonindex/#id"
  );
});
test("base/dir_1/index.md -> base/dir_1/nested/nonindex.md", () => {
  let currentFile = "base/dir_1/index.md"; // url is base/dir_1/, destination is base/dir_1/nested/nonindex/
  assert.equal(
    transformLink("./nested/nonindex.md", currentFile.endsWith("/index.md")),
    "nested/nonindex/"
  );
  assert.equal(
    transformLink("nested/nonindex.md", currentFile.endsWith("/index.md")),
    "nested/nonindex/"
  );
  assert.equal(
    transformLink("./nested/nonindex.md#id", currentFile.endsWith("/index.md")),
    "nested/nonindex/#id"
  );
  assert.equal(
    transformLink("nested/nonindex.md#id", currentFile.endsWith("/index.md")),
    "nested/nonindex/#id"
  );
});
test("base/dir_1/nonindex.md -> base/dir_1/nested/nonindex.md", () => {
  let currentFile = "base/dir_1/nonindex.md"; // url is base/dir_1/nonindex/, destination is base/dir_1/nested/nonindex/
  assert.equal(
    transformLink("./nested/nonindex.md", currentFile.endsWith("/index.md")),
    "../nested/nonindex/"
  );
  assert.equal(
    transformLink("nested/nonindex.md", currentFile.endsWith("/index.md")),
    "../nested/nonindex/"
  );
  assert.equal(
    transformLink("./nested/nonindex.md#id", currentFile.endsWith("/index.md")),
    "../nested/nonindex/#id"
  );
  assert.equal(
    transformLink("nested/nonindex.md#id", currentFile.endsWith("/index.md")),
    "../nested/nonindex/#id"
  );
});
test("base/dir_1/nonindex.md -> base/dir_1/nested/index.md", () => {
  let currentFile = "base/dir_1/nonindex.md"; // url is base/dir_1/nonindex/, destination is base/dir_1/nested/
  assert.equal(
    transformLink("./nested/index.md", currentFile.endsWith("/index.md")),
    "../nested/"
  );
  assert.equal(
    transformLink("nested/index.md", currentFile.endsWith("/index.md")),
    "../nested/"
  );
  assert.equal(
    transformLink("./nested/index.md#id", currentFile.endsWith("/index.md")),
    "../nested/#id"
  );
  assert.equal(
    transformLink("nested/index.md#id", currentFile.endsWith("/index.md")),
    "../nested/#id"
  );
});
test("base/dir_1/index.md -> base/dir_1/nested/index.md", () => {
  let currentFile = "base/dir_1/index.md"; // url is base/dir_1/, destination is base/dir_1/nested/
  assert.equal(
    transformLink("./nested/index.md", currentFile.endsWith("/index.md")),
    "nested/"
  );
  assert.equal(
    transformLink("nested/index.md", currentFile.endsWith("/index.md")),
    "nested/"
  );
  assert.equal(
    transformLink("./nested/index.md#id", currentFile.endsWith("/index.md")),
    "nested/#id"
  );
  assert.equal(
    transformLink("nested/index.md#id", currentFile.endsWith("/index.md")),
    "nested/#id"
  );
});
test("base/dir_1/index.md -> base/dir_2/index.md", () => {
  let currentFile = "base/dir_1/index.md"; // url is base/dir_1/, destination is base/dir_2/
  assert.equal(
    transformLink("../dir_2/index.md", currentFile.endsWith("/index.md")),
    "../dir_2/"
  );
  assert.equal(
    transformLink("../dir_2/index.md#id", currentFile.endsWith("/index.md")),
    "../dir_2/#id"
  );
});
test("base/dir_1/index.md -> base/index.md", () => {
  let currentFile = "base/dir_1/index.md"; // url is base/dir_1/, destination is base/
  assert.equal(
    transformLink("../../index.md", currentFile.endsWith("/index.md")),
    "../../"
  );
  assert.equal(
    transformLink("../../index.md#id", currentFile.endsWith("/index.md")),
    "../../#id"
  );
});
test("base/dir_1/index.md -> base/dir_2/nonindex.md", () => {
  let currentFile = "base/dir_1/index.md"; // url is base/dir_1/, destination is base/dir_2/nonindex/
  assert.equal(
    transformLink("../dir_2/nonindex.md", currentFile.endsWith("/index.md")),
    "../dir_2/nonindex/"
  );
  assert.equal(
    transformLink("../dir_2/nonindex.md#id", currentFile.endsWith("/index.md")),
    "../dir_2/nonindex/#id"
  );
});
test("non-md remains untouched", () => {
  let currentFile = "base/dir_1/nonindex.md"; // url is base/dir_1/nonindex/, destination is base/dir_1/nonindex_2/
  assert.equal(
    transformLink("./image.jpg", currentFile.endsWith("/index.md")),
    "./image.jpg"
  );
  assert.equal(
    transformLink("image.jpg", currentFile.endsWith("/index.md")),
    "image.jpg"
  );
  assert.equal(
    transformLink(
      "./image.jpg#github-light-mode",
      currentFile.endsWith("/index.md")
    ),
    "./image.jpg#github-light-mode"
  );
  assert.equal(
    transformLink(
      "image.jpg#github-light-mode",
      currentFile.endsWith("/index.md")
    ),
    "image.jpg#github-light-mode"
  );
});
test("full link remains untouched", () => {
  let currentFile = "base/dir_1/nonindex.md"; // url is base/dir_1/nonindex/, destination is base/dir_1/nonindex_2/
  assert.equal(
    transformLink(
      "https://somewhere.com/notes.md",
      currentFile.endsWith("/index.md")
    ),
    "https://somewhere.com/notes.md"
  );
  assert.equal(
    transformLink(
      "http://somewhere.com/notes.md",
      currentFile.endsWith("/index.md")
    ),
    "http://somewhere.com/notes.md"
  );
});
// Not sure why you would do this?
test("base/dir_1/index.md -> base/dir_1/index.md", () => {
  let currentFile = "base/dir_1/index.md";
  assert.equal(
    transformLink("./index.md", currentFile.endsWith("/index.md")),
    "#"
  );
  assert.equal(
    transformLink("./index.md#id", currentFile.endsWith("/index.md")),
    "#id"
  );
});
```

lorenzolewis commented:
Would love to include those tests in the repo. Maybe you could move this new plugin from `src/plugins/remark-rewrite-markdown-links.ts` to `src/plugins/remark-rewrite-markdown-links/index.ts` and the include a `tests.ts` file in that folder? Would also add a new command script to `package.json` to execute the test.

lorenzolewis commented:
@Aierie no rush on this, but give this a rebase when you have a moment (and lmk about adding in the tests). I put in a few changes that affect the rendering of index pages and curious if it all still works as expected with this PR.

Aierie committed:
also handle mdx file links

Aierie committed:
handle definitions too

Aierie commented:
Just tried - the plugin still works. The url trailing slash is important for the relative links to work reliably though, and locally we can't guarantee that we have trailing slashes - though on Netlify the pretty urls are adding it on.

Aierie commented:
Will start on the tests in a bit

Aierie committed:
update node types to 18
- since we specify node 18 in node-version
- needed in order to use node built in test runner

Aierie committed:
introduce tests

Aierie committed:
run mdx variant of tests

Aierie commented:
Command to run tests is `npx ts-node --esm src/plugins/remark-rewrite-markdown-links/test.ts`

Aierie committed:
types/node should be a dev dependency

lorenzolewis commented:
Tested this one and a couple of things
1. Tests look awesome! Could you add the command you mentioned as a `test` script directly in `package.json`? I know it's specific to this one test for now but if the script is there we can at least expand on it in the future
2. Removing .md (and .mdx) works like a charm! 
3. You noted for fixing the inconsistencies for links that it's assuming a trailing slash. Is there a way that it'd be able to handle either way dynamically?

For 3 above if you want to test behaviour locally you can rebase and then fire up Astro to take a look at `http://localhost:3000/api/core-js/1`. The dev server automatically handles both with and without trailing slash in the current configuration. I haven't gotten around to getting the layout on that page setup so be warned of a bright white screen beforehand 👻

lorenzolewis commented:
Btw, just noticed you don't have commit signing setup. Could you please set that up so we can have those pretty green checkmarks next to your commits? <3 

https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

Aierie committed:
add test command

Aierie commented:
I've set up commit signing and added the test command. 

re: the setup for local development, I think being strict and [adding `trailingSlash: 'always'`](https://docs.astro.build/en/reference/configuration-reference/#trailingslash) might be better because it will force anyone who's developing to always assume trailing slashes, which we should typically have in production.

Aierie commented:
I think it's hard to dynamically handle both links without and with a trailing slash if we keep relative links, unless we have a script that modifies links based on the current url, maybe? 

This problem goes away if we make this plugin generate absolute links... but that seems tricky if we ever do slug manipulation/move things around with content collections.

lorenzolewis commented:
@Aierie that makes sense to set the trailing slash to be always. I have Netlify setup to do pretty URLs (to enable trailing slash) so should match that. Could you just add that to this PR and then can run with it?

lorenzolewis commented:
Closing in favour of https://github.com/tauri-apps/tauri-docs/pull/1245 as I had to do a bit of fixing because of an old package and couldn't get this PR to modify locally. But used all the same logic you had in place @Aierie so thank you! 🧡
